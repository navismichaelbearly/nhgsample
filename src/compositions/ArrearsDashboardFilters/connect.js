import { connect } from 'react-redux';

import {
  getStatuses as getArrearStatuses,
  updateFiltersAndFetchResults,
} from '../../ducks/arrears';
import {
  getTasksSummary,
  getStatuses as getTaskStatuses,
  getMyTasksStats,
  getUserTasksSummary,
  getSummaryTasks,
} from '../../ducks/tasks';
import getHelpers from '../../util/stateHelpers';
import { sortPatchList, flattenPatches, flattenUsers } from '../../util/patch';

export const mapStateToProps = state => {
  const { get, getString } = getHelpers(state);

  const profile = get(['user', 'profile']);

  return {
    taskStatuses: get(['tasks', 'statuses']),
    arrearStatuses: get(['arrears', 'statuses']),
    arrearsDashboardLabels: getString(['arrearsDashboard']),
    filterLabels: get(['dictionary', 'arrearsDashboard', 'filters']),
    heading: getString(['arrearsDashboard', 'heading']),
    loadingArrears: get(['arrears', 'loadingArrears']),
    loadingTasks: get(['tasks', 'loading']),
    patches: sortPatchList(get(['patch', 'patchList']), profile),
    myTasks: get(['tasks', 'mySummaryItems']),
    myTasksStats: get(['tasks', 'myTasksStats']),
    mySummaryItemsCount: get(['tasks', 'mySummaryItems', 'count'], 0),
    selectedUserTasksCount: get(['tasks', 'userItemsCount'], 0),
    profile,
    youText: get(['patchList', 'you']),
  };
};

export const mergeProps = (
  { breadcrumb, heading, patches, profile, ...stateProps },
  { dispatch },
  ownProps
) => {
  const getSummary = (opts, isMine /* used when querying for 'My tasks' */) =>
    dispatch(
      getTasksSummary(
        {
          statuses: 'open',
          EntityType: 'Arrears',
          ...opts,
        },
        isMine
      )
    );
  const flattedUsers = flattenUsers(patches)
    ? flattenUsers(patches)
        .split(',')
        .filter(e => e !== profile.id)
    : [];
  const params = {
    owners: flattedUsers.toString(),
    statuses: 'open',
    EntityType: 'Arrears',
    pageSize: 6,
  };
  const flattenPatchesValues = flattenPatches(patches);
  return {
    ...stateProps,
    patches,
    profile,
    getArrearsSummary: (opts, item) => {
      dispatch(updateFiltersAndFetchResults({ patches: flattenPatchesValues, ...opts }, item));
    },
    getArrearStatuses: opts => {
      dispatch(getArrearStatuses({ patches: flattenPatchesValues, ...opts }));
    },
    getTaskStatuses: opts =>
      dispatch(getTaskStatuses({ patches: flattenPatchesValues, EntityType: 'Arrears', ...opts })),
    getSummaryTasks: opts => {
      dispatch(getSummaryTasks({ patches: flattenPatchesValues, EntityType: 'Arrears', ...opts }));
    },
    getMyTasksStats: opts => {
      dispatch(getMyTasksStats({ EntityType: 'Arrears', ...opts }));
    },
    getMyTasksSummary: () => getSummary({ userId: profile.id }),
    getSelectedUserTasks: () => {
      dispatch(getUserTasksSummary({ ...params }));
    },
    getTasksSummary: (opts = {}, isMine) => {
      getSummary({ patches: flattenPatchesValues, ...opts }, isMine);
    },
    ...ownProps,
  };
};

export default connect(mapStateToProps, null, mergeProps);
