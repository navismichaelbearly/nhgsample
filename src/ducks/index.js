import { combineReducers, createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { enableReduxDevtool } from 'nhh-styles';

import activities from './activities';
import arrears from './arrears';
import sendCorrespondence from './sendCorrespondence';
import customer from './customer';
import dictionary from './dictionary';
import features from './features';
import forms from './forms';
import housingofficersearch from './housingofficersearch';
import interactions from './interactions';
import legalReferral from './legalReferral';
import linkedCases from './linkedCases';
import media from './media';
import modal from './modal';
import navigation from './navigation';
import notes from './notes';
import notifications from './notifications';
import patch from './patch';
import pause from './pause';
import paymentPlan from './paymentPlan';
import referral from './referral';
import ribbon from './ribbon';
import tasks from './tasks';
import tenancy from './tenancy';
import user from './user';
import welfareReform from './welfareReform';

export const reducers = combineReducers({
  activities,
  arrears,
  sendCorrespondence,
  customer,
  dictionary,
  features,
  forms,
  housingofficersearch,
  interactions,
  legalReferral,
  linkedCases,
  media,
  modal,
  navigation,
  notes,
  notifications,
  patch,
  pause,
  paymentPlan,
  referral,
  ribbon,
  tasks,
  tenancy,
  user,
  welfareReform,
});

const storeEnhancer = compose(applyMiddleware(thunk), enableReduxDevtool);

export default createStore(reducers, storeEnhancer);
