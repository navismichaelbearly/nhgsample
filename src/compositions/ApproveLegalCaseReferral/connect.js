import { connect } from 'react-redux';
import path from 'ramda/src/path';

import { addNotification } from '../../ducks/notifications';
import { updateRibbon } from '../../ducks/ribbon';
import {
  cancelLegalReferral,
  getLegalReferral,
  setLegalReferralApproved,
} from '../../ducks/legalReferral';
import getHelpers from '../../util/stateHelpers';
import downloadFile from '../../util/downloadFile';
import { isUserSeniorManager } from '../../util/userRoles';
import { status as legalReferralCaseStatus } from '../../constants/legalReferral';

export const mapStateToProps = (state, ownProps) => {
  const { get, getBreadcrumbs } = getHelpers(state);
  const {
    arrearsDetails: { heading: arrearsDetailsHeading },
    approvelegalCaseReferral: {
      errorText,
      heading,
      labels,
      notification,
      subtitle,
      requiresNonCreatorApproval,
    },
  } = state.dictionary;

  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);
  const submissionId = path(['match', 'params', 'submissionId'], ownProps);
  const redirectToDetailsPage = () => ownProps.history.push(`/arrears-details/${arrearsId}`);

  const isLegalReferralCreatedByLoggedUser =
    get(['legalReferral', 'detail', 'owner', 'id']) === get(['user', 'profile', 'id']);
  const isLoggedUserManagerOrSeniorManager = get(['user', 'userType']) === 'manager';

  const isLoggedUserSeniorManager = isUserSeniorManager(get(['user', 'roles']));
  const isLegalReferralWaitingForManagerSignOff =
    get(['legalReferral', 'detail', 'progressStatus']) ===
    legalReferralCaseStatus.pendingSubmission;

  const canApproveReferral =
    !isLegalReferralCreatedByLoggedUser &&
    isLoggedUserManagerOrSeniorManager &&
    !(isLegalReferralWaitingForManagerSignOff && isLoggedUserSeniorManager);
  const showReadOnlyView = !canApproveReferral;
  const showReadOnlyViewNotification = showReadOnlyView && isLegalReferralCreatedByLoggedUser;

  return {
    activePause: get(['pause', 'activePause']),
    arrearsId,
    breadcrumb: getBreadcrumbs(
      heading,
      {},
      {
        label: arrearsDetailsHeading,
        to: `/arrears-details/${arrearsId}`,
      }
    ),
    errorText,
    formError: get(['pause', 'error']),
    heading,
    labels,
    loading: get(['legalReferral', 'loading']),
    notification,
    referralPack: get(['legalReferral', 'detail', 'referralPack']),
    submissionId,
    subtitle,
    redirectToDetailsPage,
    requiresNonCreatorApproval,
    showReadOnlyView,
    showReadOnlyViewNotification,
  };
};

export const mergeProps = (
  {
    breadcrumb,
    heading,
    notification: { approvedTitle, cancelledTitle },
    submissionId,
    redirectToDetailsPage,
    ...stateProps
  },
  { dispatch },
  ownProps
) => ({
  ...stateProps,
  getLegalReferral: () => dispatch(getLegalReferral(submissionId)),
  heading,
  onBack: () => redirectToDetailsPage(),
  onSubmit: noteText =>
    dispatch(
      noteText
        ? cancelLegalReferral(submissionId, { noteText })
        : setLegalReferralApproved(submissionId)
    )
      .then(() => {
        redirectToDetailsPage();
        dispatch(
          addNotification({
            dataBddPrefix: 'approveLegalReferral',
            icon: 'success',
            notificationType: 'confirmation',
            title: noteText ? cancelledTitle : approvedTitle,
          })
        );
      })
      .catch(() => {}),
  updatePageHeader: () => dispatch(updateRibbon({ title: heading, breadcrumb })),
  downloadFile: (uri, filename) => downloadFile(uri, filename),
  ...ownProps,
});

export default connect(mapStateToProps, null, mergeProps);
