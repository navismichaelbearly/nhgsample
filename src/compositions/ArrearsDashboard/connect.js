import clone from 'ramda/src/clone';
import { connect } from 'react-redux';
import isEmpty from 'ramda/src/isEmpty';

import { getArrearsSummary } from '../../ducks/arrears';
import {
  invalidateSummaryTasks,
  getSummaryTasks,
  getTasksSummary,
  getUserTasksSummary,
} from '../../ducks/tasks';
import getHelpers from '../../util/stateHelpers';
import { updateRibbon } from '../../ducks/ribbon';
import { showPatchSelect } from '../../ducks/patch';
import { sortPatchList, flattenPatches, flattenUsers } from '../../util/patch';

export const mapStateToProps = state => {
  const { get, getBreadcrumbs, getString } = getHelpers(state);

  const profile = get(['user', 'profile']);

  return {
    arrears: get(['arrears', 'items']),
    breadcrumb: getBreadcrumbs(),
    cardLabels: getString(['arrearsSummaryCards']),
    filterLabel: get(['arrears', 'filterLabel']),
    tasksNoDataMsg: getString(['tasksNoDataMsg']),
    myTasksNoDataMsg: getString(['myTasksNoDataMsg']),
    selectedUserTasksNoDataMsg: getString(['selectedUserTasksNoDataMsg']),
    arrearsNoDataMsg: getString(['arrearsNoDataMsg']),
    taskCardLabels: getString(['taskSummaryCards']),
    heading: getString(['arrearsDashboard', 'heading']),
    loadingArrears: get(['arrears', 'loadingArrears']),
    loadingTasks: get(['tasks', 'loading']),
    viewMoreLoading: get(['tasks', 'viewMoreLoading']),
    patches: clone(sortPatchList(get(['patch', 'patchList']), profile)),
    profile,
    resultsCount: get(['arrears', 'resultsCount']),
    selectedUserTasksCount: get(['tasks', 'userItemsCount'], 0),
    youText: getString(['patchSelect', 'you']),
    selectedUserTasks: get(['tasks', 'userItems'], []),
    myTasks: get(['tasks', 'mySummaryItems']),
    tasks: get(['tasks', 'summaryItems']),
  };
};

export const mergeProps = (
  { breadcrumb, heading, patches, profile, ...stateProps },
  { dispatch },
  ownProps
) => {
  const getSummary = (opts, isMine /* used when querying for 'My tasks' */) =>
    dispatch(
      getTasksSummary(
        {
          statuses: 'open',
          EntityType: 'Arrears',
          ...opts,
        },
        isMine
      )
    );
  const flattedUsers = !isEmpty(flattenUsers(patches))
    ? flattenUsers(patches)
        .split(',')
        .filter(e => e !== profile.id)
    : [];
  const userParams = {
    owners: flattedUsers.toString(),
    statuses: 'open',
    EntityType: 'Arrears',
    pageSize: 6,
  };
  const flattenPatchesResults = flattenPatches(patches);
  const patchParams = {
    Patches: flattenPatchesResults,
    statuses: 'open',
    EntityType: 'Arrears',
    pageSize: 6,
  };
  return {
    ...stateProps,
    profile,
    patches,
    getArrearsSummary: opts =>
      dispatch(getArrearsSummary({ patches: flattenPatchesResults, ...opts })),
    getTasksSummary: opts => dispatch(getTasksSummary({ ...opts })),
    getUserTasksSummary: () => {
      dispatch(getUserTasksSummary({ ...userParams }));
    },
    getMyTasksSummary: () => getSummary({ userId: profile.id }, true),
    getSelectedUserTasks: opts => {
      const args = { ...userParams, ...opts };
      dispatch(getUserTasksSummary({ ...args }));
    },
    getSummaryTasks: opts => {
      const args = { ...patchParams, ...opts };
      dispatch(getSummaryTasks({ ...args }));
    },
    onOpenPatchSelect: () => dispatch(showPatchSelect()),
    invalidateTasks: () => dispatch(invalidateSummaryTasks()),
    updatePageHeader: () => dispatch(updateRibbon({ title: heading, breadcrumb })),
    ...ownProps,
  };
};

export default connect(mapStateToProps, null, mergeProps);
