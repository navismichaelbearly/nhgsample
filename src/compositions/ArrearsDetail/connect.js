import { connect } from 'react-redux';
import path from 'ramda/src/path';
import moment from 'moment';
import pipe from 'ramda/src/pipe';
import prop from 'ramda/src/prop';
import flip from 'ramda/src/flip';
import find from 'ramda/src/find';
import whereEq from 'ramda/src/whereEq';
import where from 'ramda/src/where';
import evolve from 'ramda/src/evolve';
import filter from 'ramda/src/filter';
import { sortBy } from 'lodash';
import contains from 'ramda/src/contains';
import { tap } from 'nhh-styles';
import qs from 'qs';

import { employeeDashboard } from '../../constants/routes';
import getHelpers from '../../util/stateHelpers';
import { updateRibbon } from '../../ducks/ribbon';
import { addNotification, removeNotification } from '../../ducks/notifications';
import { getNotesByCaseId, invalidateNotes } from '../../ducks/notes';
import { getFilesByCaseId, invalidateDownloadedFiles } from '../../ducks/media';
import { setModalContent } from '../../ducks/modal';
import { getAccountsTransactions } from '../../ducks/customer';
import { getWelfareReform } from '../../ducks/welfareReform';
import { loadLinkedCases } from '../../ducks/linkedCases';
import downloadFile from '../../util/downloadFile';

const RENT_ACCOUNT_TYPE_CODE = 'REN';
const CHARGE_CODES_TO_SHOW = ['SERV', 'FPARK', 'CPPK'];

const getFrom = flip(prop);
const isIncludedIn = flip(contains);

export const findRentAccount = frequencies =>
  pipe(
    path(['customer', 'accounts']),
    find(whereEq({ typeCode: RENT_ACCOUNT_TYPE_CODE })),
    evolve({
      paymentFrequency: getFrom(frequencies),
      charges: filter(where({ chargeCode: isIncludedIn(CHARGE_CODES_TO_SHOW) })),
    })
  );

export const mapStateToProps = (state, ownProps) => {
  const { get, getBreadcrumbs, pickFrom, getString } = getHelpers(state);
  const {
    arrearsDetails: {
      documents,
      frequencies,
      arrearsDetailsLabels,
      heading,
      informationHeading,
      legalActions,
      linkedItemsButton,
      phasesHeading,
    },
    notes: notesText,
    paymentPlan: { installmentArrangementFormat },
  } = state.dictionary;

  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);
  const leadTenant = get(['tenancy', 'tenants', 0], {});
  const linkedCases = get(['linkedCases', 'linkedCases'], undefined);
  const linkedCaseIds = linkedCases ? sortBy(linkedCases, [t => t.createdOn]).reverse() : undefined;
  const tenancyId = get(['tenancy', 'tenancyReferenceNumber']);
  const welfareReform = get(['welfareReform', 'welfareReform']);
  const hasAccounts = !!get(['customer', 'accounts']);
  const genericErrorText = getString(['genericErrorText']);
  const queryParams = qs.parse(ownProps.location.search, { ignoreQueryPrefix: true });
  const redirectToDetailsPage = () => ownProps.history.push(`/arrears-details/${arrearsId}`);
  return {
    aModalIsVisible: get(['modal', 'visible']),
    arrears: get(['arrears', 'items']),
    arrearsDetailsLabels: {
      installmentArrangementFormat,
      ...arrearsDetailsLabels,
    },
    arrearsId,
    breadcrumb: getBreadcrumbs(heading),
    documentLabels: documents,
    documents: pickFrom(['downloadedFiles', 'downloadError', 'loading', 'loaded'], 'media'),
    genericErrorText,
    heading,
    informationHeading,
    tenancyId,
    welfareReform,
    legalActions,
    linkedCaseIds,
    loadTenancyTransactions: !!Object.keys(leadTenant).length,
    modalToShow: queryParams.modal,
    notes: get(['notes']),
    partyId: leadTenant.partyIdentifier,
    redirectToDetailsPage,
    linkedItemsButton: leadTenant.contactId
      ? {
          text: linkedItemsButton,
          href: `${employeeDashboard}/customer/${leadTenant.contactId}`,
        }
      : null,
    notesText,
    phasesHeading,
    rentDetail: tap(hasAccounts && findRentAccount(frequencies)(state), rentAccount => ({
      ...rentAccount,
    })),
  };
};

export const mergeProps = (
  { arrears, arrearsId, tenancyId, breadcrumb, notes, partyId, heading, documents, ...stateProps },
  { dispatch },
  ownProps
) => {
  const arrearsDetail = arrears.find(i => i.id === arrearsId);

  return {
    ...stateProps,
    arrearsDetail,
    arrearsId,
    partyId,
    tenancyId,
    closeTemplate: () => dispatch(setModalContent()),
    documents: {
      ...documents,
      downloadFile: (uri, filename) => downloadFile(uri, filename),
      invalidateList: () => dispatch(invalidateDownloadedFiles()),
      getList: () => dispatch(getFilesByCaseId(arrearsId)),
    },
    updatePageHeader: () => dispatch(updateRibbon({ title: heading, breadcrumb })),
    addNotification: props => dispatch(addNotification(props)),
    notes: {
      ...notes,
      invalidateList: () => dispatch(invalidateNotes()),
      getList: () => dispatch(getNotesByCaseId(arrearsId)),
    },
    openTemplate: template => {
      dispatch(setModalContent(template));
    },
    removeNotification: notificationType => dispatch(removeNotification(notificationType)),
    getTenancyTransactions: () => {
      const startDate = moment.utc().subtract(12, 'month').format('YYYY-MM-DD');
      dispatch(
        getAccountsTransactions(partyId, { startDate }, arrearsDetail.paymentReferenceNumber)
      ).catch(() => {});
    },
    loadLinkedCases: () => dispatch(loadLinkedCases(arrearsId)),
    getWelfareReformDetails: () =>
      tenancyId ? dispatch(getWelfareReform(tenancyId)).catch(() => {}) : null,
    ...ownProps,
  };
};

export default connect(mapStateToProps, null, mergeProps);
