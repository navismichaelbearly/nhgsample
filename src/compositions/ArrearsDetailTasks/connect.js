import { connect } from 'react-redux';
import path from 'ramda/src/path';
import pickAll from 'ramda/src/pickAll';

import { setModalContent } from '../../ducks/modal';
import { getTasks, invalidateTasks } from '../../ducks/tasks';
import getHelpers from '../../util/stateHelpers';
import { isUserSeniorManager } from '../../util/userRoles';
import { CANCELLED, COMPLETED, OPEN } from '../../constants/taskStatuses';

export const mapStateToProps = (state, ownProps) => {
  const { get } = getHelpers(state);
  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);
  const isLoading = get(['tasks', 'loading']);
  const {
    arrearsDetails: {
      tasks: { heading, labels },
    },
    archivedTasks,
  } = state.dictionary;
  const user = pickAll(['roles', 'profile'], get(['user']));

  const arrears = get(['arrears', 'items']);
  const currentArrear = arrears.find(arrear => arrear.id === arrearsId) || {};
  const canCreateLegalReferral =
    !isUserSeniorManager(user.roles) &&
    path(['legalReferralCase', 'canCreateLegalReferral'], currentArrear);
  const legalReferralActionStatus = path(['legalReferralCase', 'status'], currentArrear);
  const legalReferralId = path(['legalReferral', 'caseId'], currentArrear);

  return {
    archivedTasksLabels: archivedTasks,
    canCreateLegalReferral,
    legalReferralId,
    legalReferralActionStatus,
    arrearsId,
    isLoading,
    tasks: get(['tasks', 'items']),
    heading,
    labels,
    user,
  };
};

export const mergeProps = ({ arrearsId, ...stateProps }, { dispatch }, ownProps) => ({
  ...stateProps,
  arrearsId,
  invalidateTasks: () => dispatch(invalidateTasks()),
  openTaskModal: modalContent => {
    if (modalContent) {
      dispatch(setModalContent(modalContent));
    }
  },
  getArrearTasks: (archived, opts) =>
    dispatch(
      getTasks({
        EntityId: arrearsId,
        EntityType: 'Arrears',
        Statuses: archived ? `${CANCELLED},${COMPLETED}` : OPEN,
        skipFetchRoles: isUserSeniorManager(stateProps.user.roles),
        ...opts,
      })
    ),
  ...ownProps,
});

export default connect(mapStateToProps, null, mergeProps);
