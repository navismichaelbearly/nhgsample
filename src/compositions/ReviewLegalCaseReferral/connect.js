import { connect } from 'react-redux';
import path from 'ramda/src/path';
import pathOr from 'ramda/src/pathOr';

import getHelpers from '../../util/stateHelpers';
import { updateRibbon } from '../../ducks/ribbon';
import {
  cancelLegalReferral,
  getLegalReferral,
  setLegalReferralSubmitted,
  clear as clearState,
} from '../../ducks/legalReferral';
import { addNotification } from '../../ducks/notifications';
import downloadFile from '../../util/downloadFile';

export const mapStateToProps = (state, ownProps) => {
  const { get, getBreadcrumbs } = getHelpers(state);
  const {
    arrearsDetails: { heading: arrearsDetailsHeading },
    legalCaseReferral: { formError, legalCaseReferral, reviewReferral },
  } = state.dictionary;

  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);
  const submissionId = path(['match', 'params', 'submissionId'], ownProps);
  const redirectToDetailsPage = () => ownProps.history.push(`/arrears-details/${arrearsId}`);

  const arrearsItems = get(['arrears', 'items']);
  const arrearsDetail = arrearsItems.find(i => i.id === arrearsId);
  const requiresApproval = pathOr(true, ['legalReferralCase', 'requiresApproval'], arrearsDetail);
  const notificationSubmittedText = requiresApproval
    ? reviewReferral.notification.submittedText
    : reviewReferral.notification.submittedNonApprovalText;

  return {
    breadcrumb: getBreadcrumbs(
      legalCaseReferral,
      {},
      {
        label: arrearsDetailsHeading,
        to: `/arrears-details/${arrearsId}`,
      }
    ),
    errorStatus: get(['legalReferral', 'errorStatus']),
    formError,
    heading: legalCaseReferral,
    loading: get(['legalReferral', 'loading']),
    notification: reviewReferral.notification,
    notificationSubmittedText,
    redirectToDetailsPage,
    referralPack: get(['legalReferral', 'detail', 'referralPack']),
    requiresApproval,
    submissionId,
    text: { ...reviewReferral },
  };
};

export const mergeProps = (
  {
    breadcrumb,
    heading,
    notification,
    notificationSubmittedText,
    redirectToDetailsPage,
    submissionId,
    ...stateProps
  },
  { dispatch },
  ownProps
) => {
  const dispatchNotification = (dataBddPrefix, title, lines) => {
    dispatch(
      addNotification({
        dataBddPrefix,
        icon: 'success',
        lines,
        notificationType: 'confirmation',
        title,
      })
    );
    redirectToDetailsPage();
  };

  return {
    ...stateProps,
    onSubmit: () =>
      dispatch(setLegalReferralSubmitted(submissionId))
        .then(() => dispatchNotification('submitLegalCaseReferral', notificationSubmittedText))
        .catch(() => {}),
    onCancel: payload =>
      dispatch(cancelLegalReferral(submissionId, payload))
        .then(() =>
          dispatchNotification('cancelLegalCaseReferral', notification.cancelledText, [
            { text: notification.cancelledLine1 },
          ])
        )
        .catch(() => {}),
    updatePageHeader: () => dispatch(updateRibbon({ title: heading, breadcrumb })),
    getLegalReferral: () => dispatch(getLegalReferral(submissionId)),
    downloadFile: (uri, filename) => downloadFile(uri, filename),
    invalidateLegalReferralState: () => dispatch(clearState()),
    ...ownProps,
  };
};

export default connect(mapStateToProps, null, mergeProps);
