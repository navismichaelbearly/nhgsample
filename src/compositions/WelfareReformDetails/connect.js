import { connect } from 'react-redux';
import path from 'ramda/src/path';
import remove from 'ramda/src/remove';

import {
  getWelfareReform,
  patchWelfareReform,
  getAPAStatus,
  getUCStatus,
} from '../../ducks/welfareReform';
import { updateRibbon } from '../../ducks/ribbon';
import getHelpers from '../../util/stateHelpers';
import {
  employeeDashboardSearchResults,
  dashboardDetailsForCustomer,
} from '../../constants/routes';

export const mapStateToProps = (state, ownProps) => {
  const { get, getBreadcrumbs } = getHelpers(state);
  const {
    arrearsDetails: { heading: arrearsDetailsHeading },
    welfareReformDetailsEdit: { paymentDayError, labels, lastUpdated, welfareReformDetails },
  } = state.dictionary;
  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);
  const tenancyParam = path(['match', 'params', 'tenancyId'], ownProps);
  const customerId = path(['match', 'params', 'id'], ownProps);
  const apaStatusObject = get(['welfareReform', 'apaStatus']);
  const ucStatusObject = get(['welfareReform', 'ucStatus']);

  const apaStatuses = apaStatusObject;
  const ucArrearsPaymentStatuses = ucStatusObject;

  let redirectToDetailsPage;
  let hidePropertyInfo;
  if (arrearsId) {
    redirectToDetailsPage = () => ownProps.history.push(`/arrears-details/${arrearsId}`);
    hidePropertyInfo = false;
  } else {
    redirectToDetailsPage = () => ownProps.history.goBack();
    hidePropertyInfo = true;
  }
  const welfareReform = get(['welfareReform', 'welfareReform']);
  const tenancy = get(['tenancy', 'tenancyReferenceNumber']);
  return {
    arrearsId,
    breadcrumb: getBreadcrumbs(
      welfareReformDetails,
      {},
      tenancyParam
        ? [
            {
              label: 'Search Results',
              key: 'searchResults',
              path: employeeDashboardSearchResults,
            },
            {
              label: 'Customer Details',
              key: 'customerDetails',
              path: dashboardDetailsForCustomer(customerId),
            },
          ]
        : [
            {
              label: arrearsDetailsHeading,
              to: `/arrears-details/${arrearsId}`,
            },
          ]
    ),
    paymentDayError,
    hidePropertyInfo,
    heading: welfareReformDetails,
    labels,
    lastUpdated,
    redirectToDetailsPage,
    apaStatuses,
    ucArrearsPaymentStatuses,
    tenancyId: tenancy || tenancyParam,
    welfareReform,
  };
};

export const mergeProps = (
  {
    arrearsId,
    apaStatuses,
    breadcrumb,
    heading,
    hidePropertyInfo,
    redirectToDetailsPage,
    tenancyId,
    userId,
    ucArrearsPaymentStatuses,
    welfareReform,
    ...stateProps
  },
  { dispatch },
  ownProps
) => {
  const success = () => {
    redirectToDetailsPage();
  };
  return {
    ...stateProps,
    hidePropertyInfo,
    tenancyId,
    apaStatuses,
    ucArrearsPaymentStatuses,
    getWelfareReformDetails: () => dispatch(getWelfareReform(tenancyId)),
    getOptionSets: () => dispatch(getAPAStatus()).then(dispatch(getUCStatus())),
    welfareReform,
    onBack: () => redirectToDetailsPage(),
    onSubmit: payload => {
      const params = { ...payload };
      dispatch(patchWelfareReform(tenancyId, params)).then(() => {
        success();
      });
    },
    updatePageHeader: () =>
      dispatch(
        updateRibbon({
          title: heading,
          breadcrumb: tenancyId ? remove(1, 1, breadcrumb) : breadcrumb,
        })
      ),
    ...ownProps,
  };
};

export default connect(mapStateToProps, null, mergeProps);
