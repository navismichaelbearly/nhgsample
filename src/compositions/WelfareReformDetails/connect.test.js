import * as welfareReformActions from '../../ducks/welfareReform';
import { mapStateToProps, mergeProps } from './connect';
import state from '../../../__mocks__/state';

describe('WelfareReformDetails connector', () => {
  let mockState;
  let mapStateToPropsResult;

  beforeEach(() => {
    mockState = state();
    mapStateToPropsResult = mapStateToProps(mockState);
  });

  describe('mapStateToProps', () => {
    it('generates the correct props', () => {
      expect(mapStateToPropsResult).toMatchSnapshot();
    });
  });

  describe('mergeProps', () => {
    let dispatchProps;
    let result;
    let ownProps;
    let redirectToDetailsPage;
    let params;

    beforeEach(() => {
      dispatchProps = {
        dispatch: jest.fn(() => ({ then: cb => cb() })),
      };
      welfareReformActions.getWelfareReform = jest.fn();
      welfareReformActions.patchWelfareReform = jest.fn();
      redirectToDetailsPage = jest.fn();
      params = { '0': 'a', '1': 'b', '2': 'c', '3': '1', '4': '2', '5': '3' };
      result = mergeProps(
        {
          ...mapStateToPropsResult,
          redirectToDetailsPage,
        },
        dispatchProps,
        ownProps
      );
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('generates the correct props', () => {
      expect(result).toMatchSnapshot();
    });

    it('performs the correct actions for updatePageHeader', () => {
      result.updatePageHeader();
      expect(dispatchProps.dispatch).toHaveBeenCalled();
    });

    it('performs the correct actions for getWelfare when there is a tenancyId', () => {
      result.getWelfareReformDetails();
      expect(dispatchProps.dispatch).toHaveBeenCalled();
      expect(welfareReformActions.getWelfareReform).toHaveBeenCalledWith(
        mapStateToPropsResult.tenancyId
      );
    });

    it('performs the correct actions for onBack', () => {
      result.onBack();
      expect(redirectToDetailsPage).toHaveBeenCalled();
    });

    it('performs the correct actions for onSubmit for patch', () => {
      result.onSubmit('abc123', params);
      expect(dispatchProps.dispatch).toHaveBeenCalled();
      expect(welfareReformActions.patchWelfareReform).toHaveBeenCalledWith(
        mapStateToPropsResult.tenancyId,
        params
      );
      expect(redirectToDetailsPage).toHaveBeenCalled();
    });
  });
});
