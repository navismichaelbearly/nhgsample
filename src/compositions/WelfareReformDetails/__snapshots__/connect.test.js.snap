// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WelfareReformDetails connector mapStateToProps generates the correct props 1`] = `
Object {
  "apaStatuses": undefined,
  "arrearsId": undefined,
  "breadcrumb": Array [
    Object {
      "href": "#{Dashboards.Core.CustomerBaseURL}",
      "id": "breadcrumb-dashboard",
      "label": "Dashboard",
    },
    Object {
      "id": "breadcrumb-arrearsDashboard",
      "label": "My arrears",
      "to": "/",
    },
    Object {
      "label": "Arrears details",
      "to": "/arrears-details/undefined",
    },
    Object {
      "id": "breadcrumb-Welfare reform details",
      "label": "Welfare reform details",
    },
  ],
  "heading": "Welfare reform details",
  "hidePropertyInfo": true,
  "labels": Object {
    "backButton": "Back",
    "saveButton": "Save",
  },
  "lastUpdated": undefined,
  "paymentDayError": "Payment day should be between 1 and 31",
  "redirectToDetailsPage": [Function],
  "tenancyId": undefined,
  "ucArrearsPaymentStatuses": undefined,
  "welfareReform": Object {
    "apaStatusDescription": "abc",
    "bedroomTax": false,
    "benefitCap": false,
    "claimStartDate": "2018-12-23T13:30:24.051Z",
    "paymentDay": 0,
    "ucArrearsPaymentStatusDescription": "def",
    "universalCredit": false,
  },
}
`;

exports[`WelfareReformDetails connector mergeProps generates the correct props 1`] = `
Object {
  "apaStatuses": undefined,
  "getOptionSets": [Function],
  "getWelfareReformDetails": [Function],
  "hidePropertyInfo": true,
  "labels": Object {
    "backButton": "Back",
    "saveButton": "Save",
  },
  "lastUpdated": undefined,
  "onBack": [Function],
  "onSubmit": [Function],
  "paymentDayError": "Payment day should be between 1 and 31",
  "tenancyId": undefined,
  "ucArrearsPaymentStatuses": undefined,
  "updatePageHeader": [Function],
  "welfareReform": Object {
    "apaStatusDescription": "abc",
    "bedroomTax": false,
    "benefitCap": false,
    "claimStartDate": "2018-12-23T13:30:24.051Z",
    "paymentDay": 0,
    "ucArrearsPaymentStatusDescription": "def",
    "universalCredit": false,
  },
}
`;
